externs.js

```diff
@@ -0,0 +1,111 @@
+/**
+ * @constructor
+ * @param {(string|!NodeName)} nodeName
+ * @param {Object|undefined} attributes
+ * @param {Array<VNode>|undefined} children
+ * @return {undefined}
+ */
+function VNode(nodeName, attributes, children) {}
+
+/** @record
+ * @return {?}
+ */
+function NodeName() { }
+
+/** @type {Object} */
+NodeName.prototype.defaultProps
+
+/** @type {(string|!NodeName)} */
+VNode.prototype.nodeName;
+
+/** @type {Object|undefined} */
+VNode.prototype.attributes;
+
+/** @type {Array<VNode>|undefined} */
+VNode.prototypechildren;
+
+/** @type{boolean|string} */
+VNode.prototype.key;
+
+
+/** Base Component class, for the ES6 Class method of creating Components
+ * @constructor
+ * @param {Object} props
+ * @param {Object} context
+ * @return {undefined}
+ */
+function Component(props, context){}
+
+/** @type {?} */
+Component.prototype.prevState;
+/** @type {?} */
+Component.prototype.prevProps;
+/** @type {?} */
+Component.prototype.prevContext;
+/** @type {?} */
+Component.prototype.base;
+/** @type {?} */
+Component.prototype.nextBase;
+/** @type {?} */
+Component.prototype._parentComponent;
+/** @type {?} */
+Component.prototype._component;
+/** @type {?} */
+Component.prototype.__ref;
+/** @type {?} */
+Component.prototype.__key;
+/** @type {?} */
+Component.prototype._linkedStates;
+/** @type {?} */
+Component.prototype._renderCallbacks;
+
+/** @type {boolean} */
+Component.prototype._dirty;
+
+/** @type {Object} */
+Component.prototype.context;
+
+/** @type {Object} */
+Component.prototype.props;
+
+/** @type {Object|undefined} */
+Component.prototype.state;
+
+/**
+ *	@param {Object} nextProps
+ *	@param {Object|undefined} nextState
+ *	@param {Object} nextContext
+ *	@return {boolean} should the component re-render
+ */
+Component.prototype.shouldComponentUpdate = function(nextProps, nextState, nextContext){};
+
+/**
+ *	@param {string} key		The path to set - can be a dot-notated deep key
+ *	@param {string} [eventPath]	If set, attempts to find the new state value at a given dot-notated path within the object passed to the linkedState setter.
+ *	@return {Function} linkStateSetter(e)
```

Comment:
sounds like this should be `{function(!Event)}`? (lower-case function implies never-null for functions, and then you can include args).